examples/updated-vm-with-contracts/storage/contract.go

```diff
@@ -0,0 +1,148 @@
+		return runtime.ContractID(ids.Empty[:]), ErrUnknownAccount
+	}
+	if err != nil {
+		return runtime.ContractID(ids.Empty[:]), err
+	}
+	return runtime.ContractID(v[:ids.IDLen]), nil
+}
+
+func ContractBytesKey(contractID runtime.ContractID) (k []byte) {
+	// contractBytePrefix -> contractID = bytes
+	k = make([]byte, 0, 1+ids.IDLen)
+	k = append(k, contractBytePrefix)
+	k = append(k, contractID[:]...)
+	k, _ = keys.Encode(k, MaxKeySize)
+	return
+}
+
+func (c *ContractStateManager) SetContractBytes(ctx context.Context, contractID runtime.ContractID, contractBytes []byte) error {
+	return c.Insert(ctx, ContractBytesKey(contractID), contractBytes)
+}
+
+func (c *ContractStateManager) GetContractBytes(ctx context.Context, contractID runtime.ContractID) ([]byte, error) {
+	return c.GetValue(ctx, ContractBytesKey(contractID))
+}
+
+func GetAccountAddress(contractID runtime.ContractID, accountCreationData []byte) codec.Address {
+	return codec.CreateAddress(AccountTypeID, sha256.Sum256(append(contractID[:], accountCreationData...)))
+}
+
+func (c *ContractStateManager) NewAccountWithContract(ctx context.Context, contractID runtime.ContractID, accountCreationData []byte) (codec.Address, error) {
+	// TODO: don't we need to generate a random address here? where should we get the randomness?
+	// if we use the account creation data, someone could override previous accounts
+	account := GetAccountAddress(contractID, accountCreationData)
+
+	// check if the account already exists
+	if _, err := c.GetValue(ctx, AccountContractIDKey(account)); err == nil {
+		return codec.EmptyAddress, ErrContractExists
+	}
+
+	return account, c.SetAccountContract(ctx, account, contractID)
+}
+
+func AccountContractIDKey(account codec.Address) (k []byte) {
+	// contractPrefix -> account -> id prefix = contractID
+	k = make([]byte, 2+codec.AddressLen)
+	k = append(k, contractPrefix)
+	k = append(k, account[:]...)
+	k = append(k, contractIDPrefix)
+	k, _ = keys.Encode(k, MaxKeySize)
+	return
+}
+
+func (c *ContractStateManager) SetAccountContract(ctx context.Context, account codec.Address, contractID runtime.ContractID) error {
+	k := AccountContractIDKey(account)
+	return c.Insert(ctx, k, contractID[:])
+}
+
+type prefixedStateMutable struct {
+	inner  state.Mutable
+	prefix []byte
+}
+
+func (s *prefixedStateMutable) prefixKey(key []byte) (k []byte) {
+	k = make([]byte, len(s.prefix)+len(key))
+	copy(k, s.prefix)
+	copy(k[len(s.prefix):], key)
+	return
+}
+
+func (s *prefixedStateMutable) GetValue(ctx context.Context, key []byte) (value []byte, err error) {
+	return s.inner.GetValue(ctx, s.prefixKey(key))
+}
+
+func (s *prefixedStateMutable) Insert(ctx context.Context, key []byte, value []byte) error {
+	return s.inner.Insert(ctx, s.prefixKey(key), value)
+}
+
+func (s *prefixedStateMutable) Remove(ctx context.Context, key []byte) error {
+	return s.inner.Remove(ctx, s.prefixKey(key))
+}
+
+func accountStatePrefix(addr codec.Address) (k []byte) {
```

Comment:
no, I believe the `runtime` package doesn't make any assumptions about state layout. I think we could provide a default state layout since this is _third_ place I have re-written the contract manager interface

Reply:
Yeah, this needs to be more generic, I think. Can have a discussion about this offline.